homeassistant/components/surepetcare/__init__.py

```diff
@@ -0,0 +1,176 @@
+"""Support for Sure Petcare cat/pet flaps."""
+import logging
+
+from surepy import (
+    SurePetcare,
+    SurePetcareAuthenticationError,
+    SurePetcareConnectionError,
+    SurePetcareError,
+)
+import voluptuous as vol
+
+from homeassistant.const import (
+    CONF_ID,
+    CONF_NAME,
+    CONF_PASSWORD,
+    CONF_SCAN_INTERVAL,
+    CONF_TYPE,
+    CONF_USERNAME,
+)
+from homeassistant.helpers import config_validation as cv
+from homeassistant.helpers.aiohttp_client import async_get_clientsession
+from homeassistant.helpers.dispatcher import async_dispatcher_send
+from homeassistant.helpers.event import async_track_time_interval
+
+from .const import (
+    CONF_DATA,
+    CONF_FLAPS,
+    CONF_HOUSEHOLD_ID,
+    CONF_PETS,
+    DATA_SURE_PETCARE,
+    DEFAULT_SCAN_INTERVAL,
+    DOMAIN,
+    SPC,
+    TOPIC_UPDATE,
+    SureThingID,
+)
+
+_LOGGER = logging.getLogger(__name__)
+
+
+FLAP_SCHEMA = vol.Schema(
+    {vol.Required(CONF_ID): cv.positive_int, vol.Required(CONF_NAME): cv.string}
+)
+
+PET_SCHEMA = vol.Schema(
+    {vol.Required(CONF_ID): cv.positive_int, vol.Required(CONF_NAME): cv.string}
+)
+
+CONFIG_SCHEMA = vol.Schema(
+    {
+        DOMAIN: vol.Schema(
+            {
+                vol.Required(CONF_USERNAME): cv.string,
+                vol.Required(CONF_PASSWORD): cv.string,
+                vol.Required(CONF_HOUSEHOLD_ID): cv.positive_int,
+                vol.Required(CONF_FLAPS): vol.All(cv.ensure_list, [FLAP_SCHEMA]),
+                vol.Required(CONF_PETS): vol.All(cv.ensure_list, [PET_SCHEMA]),
+                vol.Optional(
+                    CONF_SCAN_INTERVAL, default=DEFAULT_SCAN_INTERVAL
+                ): cv.time_period,
+            }
+        )
+    },
+    extra=vol.ALLOW_EXTRA,
+)
+
+
+async def async_setup(hass, config):
+    """Initialize the Sure Petcare component."""
+    conf = config[DOMAIN]
+
+    # update interval
+    scan_interval = conf.get(CONF_SCAN_INTERVAL, DEFAULT_SCAN_INTERVAL)
+
+    # shared data
+    hass.data[DOMAIN] = hass.data[DATA_SURE_PETCARE] = {}
+
+    # sure petcare api connection
+    try:
+        surepy = SurePetcare(
+            conf[CONF_USERNAME],
+            conf[CONF_PASSWORD],
+            conf[CONF_HOUSEHOLD_ID],
+            hass.loop,
+            async_get_clientsession(hass),
+            debug=True,
+        )
+    except SurePetcareAuthenticationError:
+        _LOGGER.error("Unable to connect to surepetcare.io: Wrong credentials!")
+        return False
+    except (SurePetcareError, SurePetcareConnectionError) as error:
+        _LOGGER.error("Unable to connect to surepetcare.io: Wrong %s!", error)
+        return False
+
+    # add flaps
+    things = [
+        {
+            CONF_NAME: flap[CONF_NAME],
+            CONF_ID: flap[CONF_ID],
+            CONF_TYPE: SureThingID.FLAP.name,
+            CONF_DATA: await surepy.get_flap_data(flap[CONF_ID]),
```

Comment:
We shouldn't fetch data here. That is done in `SurePetcareAPI`. Just create a list of dicts with name, id and type.